design-proposals/klip-1-improve-udf-interfaces.md

```diff
@@ -0,0 +1,340 @@
+| Structured UDFs       | Arbitrary data handling without intermediary transforms                |
+
+## Public APIS
+
+Some public APIs will be changed, though all will be changed in a data backwards compatible fashion:
+
+|      Improvement      |                               API change                               |
+|:---------------------:|------------------------------------------------------------------------|
+| Generics support      | UDF interface will accept inferred generics (e.g. `List<T>`)           |
+| Varargs support       | UDF interface will accept `...` in the method declarations             |
+| Complex Aggregation   | UDAFs may include `VR export(A agg)` to convert aggregated to output   |
+| Structured UDFs       | UDF/UDAF interfaces will accept the Kafka Connect `Struct` type        |
+
+## Design
+
+Each of these improvements is relatively self-contained and can be worked on independently and in
+parallel. Below are detailed designs for each:
+
+### Structured UDFs
+
+After refactoring done in [#2411](https://github.com/confluentinc/ksql/pull/2411), the type coercion 
+ensures that the validation does not unnecessarily prevent usage of `Struct`. To complete support
+for sturcutred UDFS, we also need to have some mechanism to create the output schema. To address 
+this, we can resolve the schema as part of the UDF specification, requiring users to specify the 
+struct schema inline (which will be parsed using the same ANTLR grammar).
+
+```java
+@UdfSchema
+public static final Schema SCHEMA = SchemaBuilder.struct()...build();
+
+@Udf(schema="STRUCT<val VARCHAR, length INT>")
+public Struct generate(
+    @UdfParam(schema="STRUCT<val VARCHAR>") final Struct from
+  ) {
+  return new Struct(...);
+}
+```
+
+`UdfFactory` could then load the correct UDF given the schema by matching the schema of the type 
+against various different methods. It will be possible to support multiple methods with different
+struct schemas if the name of the udf method is changed (note that the java method name is ignored
+from actual evaluation).
+
+### Generics
+
+Today, we eagerly convert UDF output types to corresponding Connect schema types at the time that we
+compile the UDFs and not the time that we use them in a `SELECT` clause. Since there is no 
+corresponding "unresolved" type in Connect, we fail this conversion. To solve this problem, we can
+infer the type from the source schema and compile the code to work with any component type (e.g.
+omit the generics from the compiled code).
+
+```java
+@Udf("returns a sublist of 'list'")
+public <T> List<T> sublist(
+    @UdfParameter final List<T> list, 
+    @UdfParameter final int start, 
+    @UdfParameter final int end) {
+  return list.subList(start, end);
+}
+```
+
+We need to ensure that the output type can be inferred. This means that we need to fail if the 
+signature is something like `<T> List<T> convert(List<String> list)`. This can be done during 
```

Comment:
Is this example accurate? The generic type in the output is indeed present in one of the parameters.

Reply:
Perhaps the wording is unclear, but I think the example stands - I only consider `List<String>` as the parameter, while `<T>` does not show up in that (we can't infer`<T>` from `List<T>`).